### PR TITLE
sql: include memory account in PreparedStatement

### PR DESCRIPTION
**Pull request by justinj**
Follow-up to #16433, makes the lifetime of allocations for a prepared
statement more accurately tie themselves to the lifetime of the prepared
statements themselves.

### CODE REVIEW

File: pkg/sql/prepare.go [hunk 1]
```diff
@@ -44,6 +45,14 @@ type PreparedStatement struct {
 	ProtocolMeta interface{} // a field for protocol implementations to hang metadata off of.
 
 	memAcc WrappableMemoryAccount
+	// constantAcc handles the allocation of various constant-folded values which
+	// are generated while planning the statement.
+	constantAcc *mon.BoundAccount
```
Comment (knz) [inline pkg/sql/prepare.go]: Why a pointer?

Comment (justinj) [inline pkg/sql/prepare.go]: I'm always a little concerned that I'm going to accidentally copy a `BoundAccount` by value, so I tend to default to pointers, but I think you're right that using a value here is better, updated.

**Comment by knz** [comment]
Code LGTM - but what motivated this? Do you envision this would fail otherwise? If so, can you add a test?

**Comment by justinj** [comment]
This change was to address the comment you made on #16433, there's a test in that one for the original problem, and I found that issue by running the Hibernate tests. The only quantitative change from that PR is that now the memory is held until the prepared statement goes away. We could add a test to assure that the memory sticks around, but then we'd need to fine-tune the size of the expression which sounds kind of fragile, unless you have a good way to handle that?

One thing that might make testing this stuff easier (which I don't *think* already exists?) would be a like, `SET crdb_internal.memory_budget=...` command, which would allow us to dynamically set the memory budget as a test requires, or if we move to a more hierarchical logic test setup that would allow us to configure that stuff on a per-test basis.

**Comment by knz** [comment]
You can use an array literal (`ARRAY[1,2,3,4...]`) to create an arbitrary large expression.

And I *love* your idea of a configurable budget for tests. I think a tunable knob in the test server settings would achieve that.

**Comment by justinj** [comment]
How do you feel about me opening a separate issue for the tunable memory budget, and I'll include a test for this in the PR for it?

**Comment by knz** [comment]
Yes of course, I wasnt suggesting to extend this PR.

**Pull request merged**